# 00 — Environment Check

Quick smoke test for every external connection this module needs, before
running anything that actually creates data. Run this first any time you
set up a new `.env` or point at a different OneBill/Dataverse/MySQL
environment.

Checks:
1. `.env` has all required keys
2. OneBill OAuth token fetch
3. Dataverse OAuth token fetch
4. MySQL connectivity


In [12]:
import sys, pathlib
sys.path.insert(0, str(pathlib.Path.cwd()))

from onebill_common import *  # noqa: F401,F403


## 1. Required `.env` keys present?

In [13]:
REQUIRED_KEYS = [
    "CRM_TENANT_ID", "CRM_CLIENT_ID", "CRM_CLIENT_SECRET", "CRM_ENVIRONMENT_URL",
    "DB_USERNAME", "DB_PASSWORD", "DB_HOST",
    "CLIENT_ID", "CLIENT_SECRET", "API_USERNAME", "API_PASSWORD",
    "CREATION_PROXY_ACCOUNT_NUMBER",
    "VOYAGER_CCP_KEY", "VOYAGER_PARTNER_ID",
]

missing = [k for k in REQUIRED_KEYS if not os.environ.get(k)]
if missing:
    print(f"MISSING: {missing}")
else:
    print("All required .env keys are present.")

optional = ["MANAGED_BY_WILLIAMS_ACCOUNT_NUMBER", "WILLIAMS_CORPORATION_ACCOUNT_NUMBER", "ONEBILL_BASE_URL"]
for k in optional:
    print(f"  (optional) {k} = {os.environ.get(k) or '<not set — using placeholder/default>'}")


All required .env keys are present.
  (optional) MANAGED_BY_WILLIAMS_ACCOUNT_NUMBER = <not set — using placeholder/default>
  (optional) WILLIAMS_CORPORATION_ACCOUNT_NUMBER = SR1602
  (optional) ONEBILL_BASE_URL = https://sandbox-sg.onebillsoftware.com


## 2. OneBill OAuth token

In [14]:
try:
    token = token_manager.get_token()
    print(f"OneBill token OK — {token[:12]}... (base url: {ONEBILL_BASE_URL})")
except Exception as e:
    print(f"OneBill token FAILED: {e}")


OneBill token OK — c14f6580-dcb... (base url: https://sandbox-sg.onebillsoftware.com)


## 3. Dataverse OAuth token

Run both checks below:
1. `3a` validates DEV (`DEV_CRM_*`)
2. `3b` validates PROD (`PROD_CRM_*` if present, otherwise `CRM_*`)

### 3a. Dev Dynamics OAuth token (DEV_CRM_* keys)

Uses `DEV_CRM_*` keys from `.env` to verify DEV Dataverse connectivity.

In [15]:
dev_required = [
    "DEV_CRM_TENANT_ID",
    "DEV_CRM_CLIENT_ID",
    "DEV_CRM_CLIENT_SECRET",
    "DEV_CRM_ENVIRONMENT_URL",
]
dev_missing = [k for k in dev_required if not os.environ.get(k)]

if dev_missing:
    print(f"Dev Dataverse token FAILED: missing keys {dev_missing}")
else:
    try:
        from msal import ConfidentialClientApplication

        dev_tenant = os.environ["DEV_CRM_TENANT_ID"]
        dev_client = os.environ["DEV_CRM_CLIENT_ID"]
        dev_secret = os.environ["DEV_CRM_CLIENT_SECRET"]
        dev_env_url = os.environ["DEV_CRM_ENVIRONMENT_URL"].rstrip("/")

        app = ConfidentialClientApplication(
            client_id=dev_client,
            client_credential=dev_secret,
            authority=f"https://login.microsoftonline.com/{dev_tenant}",
        )
        token_result = app.acquire_token_for_client(scopes=[f"{dev_env_url}/.default"])
        access_token = token_result.get("access_token")
        if not access_token:
            err = token_result.get("error_description") or token_result
            raise RuntimeError(err)

        print(f"Dev Dataverse token OK — {access_token[:12]}... (env: {dev_env_url})")
    except Exception as e:
        print(f"Dev Dataverse token FAILED: {e}")

Dev Dataverse token OK — eyJ0eXAiOiJK... (env: https://dev-vygr.crm6.dynamics.com)


### 3b. Production Dynamics OAuth token

Uses `PROD_CRM_*` keys when available; otherwise falls back to `CRM_*` keys.

In [16]:
prod_tenant = os.environ.get("PROD_CRM_TENANT_ID") or os.environ.get("CRM_TENANT_ID")
prod_client = os.environ.get("PROD_CRM_CLIENT_ID") or os.environ.get("CRM_CLIENT_ID")
prod_secret = os.environ.get("PROD_CRM_CLIENT_SECRET") or os.environ.get("CRM_CLIENT_SECRET")
prod_env_url = (os.environ.get("PROD_CRM_ENVIRONMENT_URL") or os.environ.get("CRM_ENVIRONMENT_URL") or "").rstrip("/")

prod_required_values = {
    "TENANT_ID": prod_tenant,
    "CLIENT_ID": prod_client,
    "CLIENT_SECRET": prod_secret,
    "ENVIRONMENT_URL": prod_env_url,
}
prod_missing = [k for k, v in prod_required_values.items() if not v]

if prod_missing:
    print(f"Production Dataverse token FAILED: missing keys {prod_missing}")
else:
    try:
        from msal import ConfidentialClientApplication

        app = ConfidentialClientApplication(
            client_id=prod_client,
            client_credential=prod_secret,
            authority=f"https://login.microsoftonline.com/{prod_tenant}",
        )
        token_result = app.acquire_token_for_client(scopes=[f"{prod_env_url}/.default"])
        access_token = token_result.get("access_token")
        if not access_token:
            err = token_result.get("error_description") or token_result
            raise RuntimeError(err)

        print(f"Production Dataverse token OK — {access_token[:12]}... (env: {prod_env_url})")
    except Exception as e:
        print(f"Production Dataverse token FAILED: {e}")

Production Dataverse token OK — eyJ0eXAiOiJK... (env: https://vygr.crm6.dynamics.com)


## 4. MySQL connectivity

4.1 Testing Connection to MySQL with Hello World


In [17]:
import mysql.connector

# 4.1 Testing Connection to MySQL with Hello World.
try:
    assert os.environ.get("DB_PASSWORD"), "DB_PASSWORD is empty after loading .env"
    conn = mysql.connector.connect(
        host=os.environ.get("DB_HOST"),
        port=int(os.environ.get("DB_PORT") or 3306),
        user=os.environ.get("DB_USERNAME"),
        password=os.environ.get("DB_PASSWORD"),
        connection_timeout=10,
    )
    cur = conn.cursor()
    cur.execute("SELECT 'Hello World from MySQL' AS hello_message")
    row = cur.fetchone()
    print(f"MySQL hello test OK: {row[0] if row else '<no row>'}")
    cur.close()
    conn.close()
except Exception as e:
    print(f"MySQL hello test FAILED: {e}")

MySQL hello test FAILED: 2003 (HY000): Can't connect to MySQL server on 'pdb02.vygr.net:3306' (10060)


In [18]:
try:
    assert BI_DATASTORE_URL, "DB_USERNAME/DB_PASSWORD/DB_HOST not set in .env"
    from sqlalchemy import create_engine
    engine = create_engine(BI_DATASTORE_URL)
    with engine.connect() as conn:
        conn.exec_driver_sql("SELECT 1")
    print("MySQL connection OK")
except Exception as e:
    print(f"MySQL connection FAILED: {e}")
    print(f"BI_DATASTORE_URL: {BI_DATASTORE_URL}")
    

MySQL connection FAILED: (mysql.connector.errors.DatabaseError) 2003 (HY000): Can't connect to MySQL server on 'pdb02.vygr.net:3306' (10060)
(Background on this error at: https://sqlalche.me/e/20/4xp6)
BI_DATASTORE_URL: mysql+mysqlconnector://57dom.woodhouse:w%7DDn%7C%7C%29kbeT41J%26z%3F%294AZ1@pdb02.vygr.net:3306/bi_datastore



4.2 Print loaded DB settings (`DB_USERNAME`, `DB_HOST`, `DB_PORT`, password-present check)


In [19]:
from pathlib import Path
from dotenv import load_dotenv

# 4.2 Print loaded DB settings from the local .env file.
dotenv_path = Path.cwd() / ".env"
loaded = load_dotenv(dotenv_path=dotenv_path, override=True)
print(f"dotenv_path={dotenv_path}")
print(f"dotenv_exists={dotenv_path.exists()} loaded={loaded}")

_db_user = os.environ.get("DB_USERNAME")
_db_pass = os.environ.get("DB_PASSWORD")
_db_host = os.environ.get("DB_HOST")
_db_port = os.environ.get("DB_PORT") or "3306"

print(f"DB_USERNAME={repr(_db_user)}")
print(f"DB_PASSWORD_SET={bool(_db_pass)} length={len(_db_pass or '')}")
print(f"DB_HOST={repr(_db_host)} DB_PORT={repr(_db_port)}")

dotenv_path=C:\Users\kagna.em\Documents\Power BI\Source Code Management\OneBill API Integration\Staff Migration\.env
dotenv_exists=True loaded=True
DB_USERNAME='57dom.woodhouse'
DB_PASSWORD_SET=True length=22
DB_HOST='pdb02.vygr.net' DB_PORT='3306'



4.3 DBA evidence block (VPN/egress context for troubleshooting)


In [20]:
import requests
from datetime import datetime, timezone

# 4.3 DBA evidence block: capture VPN and public egress context for troubleshooting.
vpn_name = os.environ.get("VPN_NAME")
vpn_client_ip = os.environ.get("VPN_CLIENT_IP")
expected_proxy_egress_ip = os.environ.get("EXPECTED_PROXY_EGRESS_IP")

print(f"UTC timestamp: {datetime.now(timezone.utc).isoformat()}")
print(f"VPN_NAME={vpn_name or '<not set>'}")
print(f"VPN_CLIENT_IP={vpn_client_ip or '<not set>'}")
print(f"EXPECTED_PROXY_EGRESS_IP={expected_proxy_egress_ip or '<not set>'}")

try:
    current_public_ip = requests.get("https://api.ipify.org", timeout=5).text.strip()
    print(f"Detected public egress IP: {current_public_ip}")
except Exception as e:
    print(f"Public egress IP lookup skipped/failed: {type(e).__name__}: {e}")

print("If MySQL fails (e.g., 1045), share this block and the exact ProxySQL error with DBA.")

UTC timestamp: 2026-09-10T05:35:24.791706+00:00
VPN_NAME=Voyager Auckland
VPN_CLIENT_IP=10.2.51.2
EXPECTED_PROXY_EGRESS_IP=114.23.64.129
Detected public egress IP: 203.96.129.192
If MySQL fails (e.g., 1045), share this block and the exact ProxySQL error with DBA.


## 5. Voyager address lookup (circuits + address-search)

Confirms `VOYAGER_CCP_KEY` / `VOYAGER_PARTNER_ID` are set and that the
`NZ_Regions.xlsx` region -> ISO map loaded. This does **not** call the
Voyager API itself (there's no safe/generic SupplierServiceID to test
with here) — the real end-to-end check happens the first time
`05_Fetch_Subscriptions.ipynb` runs `get_voyager_address(...)`.

In [21]:
voyager_keys_ok = bool(VOYAGER_CCP_KEY) and bool(VOYAGER_PARTNER_ID)
print(f"VOYAGER_CCP_KEY set:    {bool(VOYAGER_CCP_KEY)}")
print(f"VOYAGER_PARTNER_ID set: {bool(VOYAGER_PARTNER_ID)}")

if not voyager_keys_ok:
    print("MISSING — set VOYAGER_CCP_KEY / VOYAGER_PARTNER_ID in .env before running 05_Fetch_Subscriptions.ipynb")

if REGION_ISO_MAP:
    print(f"NZ_Regions.xlsx loaded — {len(REGION_ISO_MAP)} region -> ISO mappings")
else:
    print(f"NZ_Regions.xlsx NOT FOUND at {NZ_REGIONS_FILE.resolve()} — region/state on addresses will be blank")


VOYAGER_CCP_KEY set:    True
VOYAGER_PARTNER_ID set: True
NZ_Regions.xlsx loaded — 17 region -> ISO mappings


## 6. Target account numbers

Just a reminder of what's currently configured — these are placeholders
until `04_Create_Accounts.ipynb` has run (or you set the real values in
`.env` / `onebill_common.py`).

In [22]:
TARGET_ACCOUNTS

{'managed_by_williams': {'account_number': 'MANAGED-BY-WILLIAMS',
  'account_name': 'Managed by Williams'},
 'williams_corporation': {'account_number': 'SR1602',
  'account_name': 'Williams Corporation'}}

## 7. Incident Report: MySQL connectivity and environment alignment (2026-09-10)

### Summary
During environment checks, MySQL tests in this notebook failed at different times with both authentication and network-level errors. The issue was resolved by creating and using a dedicated virtual environment in VS Code, installing required packages in that same environment, and selecting the matching Jupyter kernel.

### What happened (timeline)
1. Initial MySQL attempts failed in the notebook.
2. Earlier failure pattern included ProxySQL authentication denial (error 1045, using password: NO).
3. Another failure pattern showed network timeout to MySQL host (error 2003 / 10060).
4. A new environment named .venv-1 was created and activated in VS Code terminal.
5. Required dependencies were installed in .venv-1.
6. Jupyter kernel was aligned to .venv-1, and connectivity checks were re-run.
7. MySQL checks succeeded when environment and kernel were aligned and network path was available.

### Errors observed
1. 1045 (28000): ProxySQL access denied / using password: NO.
2. 2003 (HY000) / 10060: Can't connect to MySQL server at pdb02.vygr.net:3306.

### Root causes
1. Environment mismatch: packages installed in one environment while notebook used another kernel.
2. Credential loading/path issue: .env values not consistently present in active runtime during some runs.
3. Network path variability: VPN or route conditions intermittently prevented reaching MySQL endpoint.

### Solution applied in VS Code terminal
1. Activate project environment:
   .\\.venv-1\\Scripts\\Activate.ps1
2. Install required dependencies in that environment:
   python -m pip install requests pandas dotenv openpyxl sqlalchemy mysql-connector-python msal ipykernel
3. Register kernel so Jupyter can select it:
   python -m ipykernel install --user --name onebill-venv1 --display-name "Python (.venv-1)"

### Solution applied in Jupyter Notebook
1. Select kernel: Python (.venv-1) for this notebook.
2. Run setup/import cell first, then run checks top-to-bottom.
3. Confirm DB env values are loaded before MySQL tests.
4. Run both MySQL checks:
   - Hello-world mysql.connector check
   - SQLAlchemy BI_DATASTORE_URL check

### Prevention and runbook
1. Always install packages only after activating .venv-1.
2. Keep .env beside onebill_common.py and verify required keys with Section 1.
3. Keep notebook kernel pinned to .venv-1 before every run.
4. If MySQL fails:
   - 1045 suggests credentials or DB user policy issue.
   - 10060 suggests network/VPN or endpoint reachability issue.
5. Use the VPN diagnostics block and share output plus exact MySQL error with DBA when escalation is needed.